### PR TITLE
update disko config, add luks encryption + initrd ssh server

### PR DESCRIPTION
**Pull request by Lassulus**

### CODE REVIEW

File: nix/modules/raid-config.nix [hunk 1]
```diff
@@ -72,9 +62,14 @@ in
         type = "mdadm";
         level = 1;
         content = {
-          type = "filesystem";
-          format = "ext4";
-          mountpoint = "/";
+          type = "luks";
+          name = "root-encrypted";
+          settings.keyFile = "/run/diskKeyFile";
```
Comment (Lassulus) [inline nix/modules/raid-config.nix:67]: this one also needs pregeneration, but only during the initial disko creation phase. the manager then needs to provide that secret during the initrd booting

File: nix/modules/hardware.nix [hunk 2]
```diff
@@ -29,5 +45,22 @@
...
     boot.loader.grub.efiInstallAsRemovable = true;
+
+    # initrd networking for unlocking LUKS
+    boot.initrd.network = {
+      enable = true;
+      ssh = {
+        enable = true;
+        authorizedKeys = config.users.extraUsers.root.openssh.authorizedKeys.keys;
+        port = 2222;
+        hostKeys = config.kuutamo.disko.keyFiles;
```
Comment (Mic92) [inline nix/modules/hardware.nix:56]: How does this work?

File: mgr/src/utils.rs [hunk 3]
```diff
@@ -22,3 +26,68 @@ pub fn timeout_ssh(host: &Host, command: &[&str], learn_known_host_key: bool) ->
...
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    let mut stdin = ssh.stdin.take().ok_or(anyhow!("could not pipe stdin"))?;
+    let mut stdout = ssh.stdout.take().ok_or(anyhow!("could not pipe stdout"))?;
+    if stdin.write_all(key.as_slice()).is_ok() {
+        let _ = stdin.write(b"\n")?;
+    } else {
+        return Err(anyhow!("fail to enter password"));
```
Comment (JosephGoulden) [inline mgr/src/utils.rs:81]:
```suggestion
        return Err(anyhow!("failed to enter password"));
```

File: mgr/src/utils.rs [hunk 4]
```diff
@@ -22,3 +26,68 @@ pub fn timeout_ssh(host: &Host, command: &[&str], learn_known_host_key: bool) ->
...
+    }
+    println!("$ ssh {}", args.join(" "));
+
+    let mut buf_string = String::new();
+
+    // Check the `cryptsetup-askpass` is asking the disk encryption key
+    if stdout.read_to_string(&mut buf_string).is_ok() && buf_string.starts_with("Passphrase for") {
+        Ok(())
+    } else {
+        Err(anyhow!("sshd response unepxected"))
```
Comment (JosephGoulden) [inline mgr/src/utils.rs:91]:
```suggestion
        Err(anyhow!("sshd response unexpected"))
```

File: nix/modules/hardware.nix [hunk 5]
```diff
@@ -15,19 +22,56 @@
...
+      ssh = {
+        enable = true;
+        port = 2222;
+        authorizedKeys = config.users.extraUsers.root.openssh.authorizedKeys.keys;
+        hostKeys = [
+          "/var/lib/secrets/sshd_key"
+        ];
+      };
+      postCommands = ''
+        ip link set dev ${config.kuutamo.disko.networkInterface} up
```
Comment (JosephGoulden) [inline nix/modules/hardware.nix:61]: Are all these ip commands needed? `boot.initrd.network.enable = true` works fine for me.

**Review by JosephGoulden — COMMENTED** [review]
Looks good, I just have those questions about the network interface.

File: nix/modules/hardware.nix [hunk 6]
```diff
@@ -15,19 +22,56 @@
       "xhci_pci"
       "ahci"
       "nvme"
+      "r8169"
+      "ixgbe"
```
Comment (Mic92) [inline nix/modules/hardware.nix:26]:
```suggestion
      "ixgbe"
      "igc" # 2.5GbitE adapter
```

File: nix/modules/hardware.nix (hunk 5 above)
Comment (Mic92) [reply nix/modules/hardware.nix:62]: Not all providers do dhcp, so this is more generic.

File: mgr/src/config.rs [hunk 7]
```diff
@@ -221,6 +221,11 @@ struct HostConfig {
     #[toml_example(default = [])]
     kld_api_ip_access_list: Vec<IpAddr>,
 
+    /// The interface to access network
+    #[serde(default)]
+    #[toml_example(default = "eth0")]
+    network_interface: Option<String>,
```
[the remark this answers is not in the document]
Comment (Mic92) [reply mgr/src/config.rs:227]: In our tests we have multiple interfaces and ovh does have at least two.

**Pull request merged**